New node `fail to write block` error when syncing near most recent block

## New node `fail to write block` error when syncing near most recent block

### Description

When a new peer is syncing, it looks like after the last batch sync of blocks, we sometimes get a failure:
```
`[ERROR] polygon.consensus.syncer: failed to write block: err="parent of 0x5dda26ea7e3157baa4fa07746be00c43c2f084a946a2fb171634650f34efbc27 (853608) not found: 0xc2cbd6ac4ed7b7c340f6070bc2e13caef6ec582e14acffca41f49f9ae50760c9"`
```

See larger log snippet:
```
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.442Z [INFO]  polygon.blockchain: write blocks: num=10 from=852347 to=852356 parent=0xcb750c6af46655acc5d8a2ef0ff555da7cace894c761da985a99a3d3862dbeb5
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.479Z [INFO]  polygon.blockchain: new block: number=852347 hash=0x12039ac1d933d9a0c4761ae84ecbb227a7e8232b2b590d0b90965037e6e9fe51 txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.480Z [INFO]  polygon.blockchain: new block: number=852348 hash=0xe550e56c73dd0695a218cb509e993b18a0c7229e0a6923b2caee328c76b60029 txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.482Z [INFO]  polygon.blockchain: new block: number=852349 hash=0xd1464576099a359200620cfa5e8c468b8efa61affe8843fef3260528a4b9f2c8 txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.482Z [INFO]  polygon.blockchain: new block: number=852350 hash=0x36c9b9114dd72fa7c3d6c6c7d78910e2acec6a038e9860c1571ea1592b072e2a txns=0 generation_time_in_sec=10
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.483Z [INFO]  polygon.blockchain: new block: number=852351 hash=0xf1f1e6c70bf3caedb625e2087b5c1df5075bd1550c6cd2c7e914462e9479933e txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.485Z [INFO]  polygon.blockchain: new block: number=852352 hash=0x2a0f4cef8616c30d40d9dcaa9d421e31545fac74d047b758ae95fe07861714f5 txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.486Z [INFO]  polygon.blockchain: new block: number=852353 hash=0x7d0b63698c309d5aab3f28d64bd0cd30ce73356c003151509ca9585b03a3c346 txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.486Z [INFO]  polygon.blockchain: new block: number=852354 hash=0x1d83f22145427ccf4861da7ebfc5374ebd03ffd004e28f42669e5ba63b481e99 txns=1 generation_time_in_sec=10
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.489Z [INFO]  polygon.blockchain: new block: number=852355 hash=0x2d3147d271012dd927da8d0fc41f9c421594abe9a86c62ddd147fbbcc61f236e txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.490Z [INFO]  polygon.blockchain: new block: number=852356 hash=0x264a473677e736779edb91cc4753b90cf368d04f85395ea913114583599ee7e8 txns=0 generation_time_in_sec=2
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.490Z [INFO]  polygon.blockchain: new head: hash=0x264a473677e736779edb91cc4753b90cf368d04f85395ea913114583599ee7e8 number=852356
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.492Z [INFO]  polygon.blockchain: write block: num=853608 parent=0xc2cbd6ac4ed7b7c340f6070bc2e13caef6ec582e14acffca41f49f9ae50760c9
Sep 29 15:27:04 ip-172-31-25-125 main[72967]: 2021-09-29T15:27:04.496Z [ERROR] polygon.consensus.syncer: failed to write block: err="parent of 0x5dda26ea7e3157baa4fa07746be00c43c2f084a946a2fb171634650f34efbc27 (853608) not found: 0xc2cbd6ac4ed7b7c340f6070bc2e13caef6ec582e14acffca41f49f9ae50760c9"
```

Restarting the node fixes the issue.

### Your environment

* OS and version `Ubuntu 20`
* version of the Polygon SDK `96be904 `
* branch that causes this issue `develop`

### Steps to reproduce

1. Have 4-node validator set writing and validating new blocks
2. Connect 5th peer and observe `polygon.blockchain: new block` output that occurs during syncing
3. When almost synced to the current block, we see the error above

### Expected behaviour

There shouldn't be an error that requires us to restart the node.

### Actual behaviour

An error stops the node from further syncing, requiring a restart.

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -120,13 +120,13 @@ func Factory(
 
 // Start starts the IBFT consensus
 func (i *Ibft) Start() error {
+	// Start the syncer
+	i.syncer.Start()
+
 	// Set up the snapshots
 	if err := i.setupSnapshot(); err != nil {
 		return err
 	}
-
-	// Start the syncer
-	i.syncer.Start()
 
 	// Start the actual IBFT protocol
 	go i.start()

--- a/protocol/service_v1.go
+++ b/protocol/service_v1.go
@@ -42,7 +42,13 @@ func (s *serviceV1) Notify(ctx context.Context, req *proto.NotifyReq) (*empty.Em
 	if err := b.UnmarshalRLP(req.Raw.Value); err != nil {
 		return nil, err
 	}
+	status, err := fromProto(req.Status)
+	if err != nil {
+		return nil, err
+	}
+
 	s.syncer.enqueueBlock(id, b)
+	s.syncer.updatePeerStatus(id, status)
 	return &empty.Empty{}, nil
 }
 

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -256,7 +256,7 @@ func (s *Syncer) updatePeerStatus(peerID peer.ID, status *Status) {
 // Broadcast broadcasts a block to all peers
 func (s *Syncer) Broadcast(b *types.Block) {
 	// diff is number in ibft
-	diff := new(big.Int).SetUint64(b.Number())
+	diff := new(big.Int).SetUint64(b.Header.Difficulty)
 
 	// broadcast the new block to all the peers
 	req := &proto.NotifyReq{

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -252,7 +252,7 @@ func (s *Syncer) enqueueBlock(peerID peer.ID, b *types.Block) {
 }
 
 func (s *Syncer) updatePeerStatus(peerID peer.ID, status *Status) {
-	s.logger.Debug("update peer status", "peer", peerID, "number", status.Number, "hash", status.Hash, "difficulty", status.Difficulty)
+	s.logger.Debug("update peer status", "peer", peerID, "latest block number", status.Number, "latest block hash", status.Hash, "difficulty", status.Difficulty)
 
 	if p, ok := s.peers[peerID]; ok {
 		p.updateStatus(status)

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -16,6 +16,7 @@ import (
 	"github.com/hashicorp/go-hclog"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
 	any "google.golang.org/protobuf/types/known/anypb"
 	empty "google.golang.org/protobuf/types/known/emptypb"
 )
@@ -42,6 +43,11 @@ func (s *syncPeer) Number() uint64 {
 	defer s.statusLock.RUnlock()
 
 	return s.status.Number
+}
+
+// IsClosed returns whether peer's connectivity has been closed
+func (s *syncPeer) IsClosed() bool {
+	return s.conn.GetState() == connectivity.Shutdown
 }
 
 // purgeBlocks purges the cache of broadcasted blocks the node has written so far
@@ -450,8 +456,12 @@ func (s *Syncer) WatchSyncWithPeer(p *syncPeer, handler func(b *types.Block) boo
 
 	// listen and enqueue the messages
 	for {
-		b := p.popBlock()
+		if p.IsClosed() {
+			s.logger.Error("Connection to a peer has closed already", "id", p.peer)
+			break
+		}
 
+		b := p.popBlock()
 		if err := s.blockchain.WriteBlocks([]*types.Block{b}); err != nil {
 			s.logger.Error("failed to write block", "err", err)
 			break

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -375,7 +375,9 @@ func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
 
 func (s *Syncer) DeleteUser(peerID peer.ID) error {
 	if p, ok := s.peers[peerID]; ok {
-		p.conn.Close()
+		if err := p.conn.Close(); err != nil {
+			return err
+		}
 		delete(s.peers, peerID)
 	}
 	return nil

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -27,11 +27,9 @@ type syncPeer struct {
 	peer   peer.ID
 	conn   *grpc.ClientConn
 	client proto.V1Client
-	status *Status
 
-	// current status
-	number uint64
-	hash   types.Hash
+	status     *Status
+	statusLock sync.RWMutex
 
 	enqueueLock sync.Mutex
 	enqueue     []*types.Block
@@ -40,7 +38,10 @@ type syncPeer struct {
 
 // Number returns the latest peer block height
 func (s *syncPeer) Number() uint64 {
-	return s.number
+	s.statusLock.RLock()
+	defer s.statusLock.RUnlock()
+
+	return s.status.Number
 }
 
 // purgeBlocks purges the cache of broadcasted blocks the node has written so far
@@ -87,14 +88,17 @@ func (s *syncPeer) appendBlock(b *types.Block) {
 	// append to the end
 	s.enqueue = append(s.enqueue, b)
 
-	// update the number and hash
-	s.number = b.Number()
-	s.hash = b.Hash()
-
 	select {
 	case s.enqueueCh <- struct{}{}:
 	default:
 	}
+}
+
+func (s *syncPeer) updateStatus(status *Status) {
+	s.statusLock.Lock()
+	defer s.statusLock.Unlock()
+
+	s.status = status
 }
 
 // Status defines the up to date information regarding the peer
@@ -121,6 +125,20 @@ func (s *Status) toProto() *proto.V1Status {
 		Hash:       s.Hash.String(),
 		Difficulty: s.Difficulty.String(),
 	}
+}
+
+// fromProto converts a proto.V1Status to a Status object
+func fromProto(status *proto.V1Status) (*Status, error) {
+	diff, ok := new(big.Int).SetString(status.Difficulty, 10)
+	if !ok {
+		return nil, fmt.Errorf("failed to parse difficulty: %s", status.Difficulty)
+	}
+
+	return &Status{
+		Number:     status.Number,
+		Hash:       types.StringToHash(status.Hash),
+		Difficulty: diff,
+	}, nil
 }
 
 // statusFromProto extracts a Status object from a passed in proto.V1Status
@@ -224,6 +242,14 @@ func (s *Syncer) enqueueBlock(peerID peer.ID, b *types.Block) {
 	p, ok := s.peers[peerID]
 	if ok {
 		p.appendBlock(b)
+	}
+}
+
+func (s *Syncer) updatePeerStatus(peerID peer.ID, status *Status) {
+	s.logger.Debug("update peer status", "peer", peerID, "number", status.Number, "hash", status.Hash, "difficulty", status.Difficulty)
+
+	if p, ok := s.peers[peerID]; ok {
+		p.updateStatus(status)
 	}
 }
 
@@ -337,8 +363,6 @@ func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
 		client:    clt,
 		status:    status,
 		enqueueCh: make(chan struct{}),
-		number:    status.Number,
-		hash:      status.Hash,
 	}
 	return nil
 }

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -459,7 +459,7 @@ func (s *Syncer) WatchSyncWithPeer(p *syncPeer, handler func(b *types.Block) boo
 	// listen and enqueue the messages
 	for {
 		if p.IsClosed() {
-			s.logger.Error("Connection to a peer has closed already", "id", p.peer)
+			s.logger.Info("Connection to a peer has closed already", "id", p.peer)
 			break
 		}
 

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -25,6 +25,7 @@ const maxEnqueueSize = 50
 // syncPeer is a representation of the peer the node is syncing with
 type syncPeer struct {
 	peer   peer.ID
+	conn   *grpc.ClientConn
 	client proto.V1Client
 	status *Status
 
@@ -276,17 +277,21 @@ func (s *Syncer) Start() {
 				return
 			}
 
-			if evnt.Type != network.PeerEventConnected {
-				continue
-			}
+			switch evnt.Type {
+			case network.PeerEventConnected:
+				stream, err := s.server.NewStream(syncerV1, evnt.PeerID)
+				if err != nil {
+					s.logger.Error("failed to open a stream", "err", err)
+					continue
+				}
+				if err := s.HandleUser(evnt.PeerID, libp2pGrpc.WrapClient(stream)); err != nil {
+					s.logger.Error("failed to handle user", "err", err)
+				}
 
-			stream, err := s.server.NewStream(syncerV1, evnt.PeerID)
-			if err != nil {
-				s.logger.Error("failed to open a stream", "err", err)
-				continue
-			}
-			if err := s.HandleUser(evnt.PeerID, libp2pGrpc.WrapClient(stream)); err != nil {
-				s.logger.Error("failed to handle user", "err", err)
+			case network.PeerEventDisconnected:
+				if err := s.DeleteUser(evnt.PeerID); err != nil {
+					s.logger.Error("failed to delete user", "err", err)
+				}
 			}
 		}
 	}()
@@ -328,9 +333,20 @@ func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
 	}
 	s.peers[peerID] = &syncPeer{
 		peer:      peerID,
+		conn:      conn,
 		client:    clt,
 		status:    status,
 		enqueueCh: make(chan struct{}),
+		number:    status.Number,
+		hash:      status.Hash,
+	}
+	return nil
+}
+
+func (s *Syncer) DeleteUser(peerID peer.ID) error {
+	if p, ok := s.peers[peerID]; ok {
+		p.conn.Close()
+		delete(s.peers, peerID)
 	}
 	return nil
 }

